neural_compressor/adaptor/torch_utils/waq/smooth_quant.py

```diff
@@ -0,0 +1,602 @@
+            self.percentile = percentile
+            self.op_types = op_types
+            self.scales_per_op = scales_per_op
+            self.calib_iter = calib_iter
+            return need_calib
+
+        if (
+            self.percentile == percentile
+            and self.op_types == op_types
+            and self.scales_per_op == scales_per_op
+            and self.calib_iter == calib_iter
+        ):
+            if isinstance(alpha, float) or self.alpha == "auto":
+                need_calib = False
+
+        self.alpha, self.percentile = alpha, percentile
+        self.op_types, self.scales_per_op = op_types, scales_per_op
+        self.calib_iter = calib_iter
+        return need_calib
+
+    def _get_sq_layer_names(self):
+        """Get the all the hook sq layer
+        :return: All the sq layer names."""
+        ##TODO this may not fit for folding=False
+        module_names = []
+        for key in self.absorb_to_layer:
+            module_names += self.absorb_to_layer[key]
+        return module_names
+
+    @torch.no_grad()
+    def _parse_absorb_to_layers(self, op_types, folding):
+        str_op_types = [i.__name__ for i in op_types]
+
+        self_absorb_layers = {}
+        if self.insert_mul:
+            self_absorb_layers = self._get_all_layer_names(op_types)  # TODO: only support linear now.
+            # fetch modules with the same input
+            group_modules = self._trace(str_op_types, skip_unsupported_layers=False)
+            if group_modules is not None:
+                # use one input for qkv
+                for k, v in group_modules.items():
+                    for i in v:
+                        if i in self_absorb_layers:
+                            self_absorb_layers.pop(i)
+                    self_absorb_layers[v[0]] = v
+                logger.debug(f"self_absorb_layers:{self_absorb_layers}")
+        if self.allow_absorb:
+            self.absorb_to_layer, no_absorb_layers = self._trace(str_op_types)
+            if self.absorb_to_layer is None and no_absorb_layers is None:
+                return None
+
+        # remove self.self_absorb_layers if it exists in self.absorb_to_layer
+        for k, v in self.absorb_to_layer.items():
+            for i in v:
+                if i in self_absorb_layers:
+                    self_absorb_layers.pop(i)
+        self.absorb_to_layer.update(self_absorb_layers)
+
+        if self.absorb_to_layer is None and no_absorb_layers is None:
+            logger.warning(
+                "sorry, could not trace the model, smooth quant is ignored."
+                "If you are using huggingface model,"
+                "you could set torchscript to True "
+            )
+            return None
+
+        # Check if input_maxes match self.absorb_to_layer
+        # (due to self._get_all_layer_names use layer tree instead of forward_path)
+        if not folding:
+            if len(self.input_mins) == 0:  ##there are some modules not used in forward
```

Comment:
@yintong-lu  For some models like bridgetower or moe model, the activate module of one example will not be align with the modules in models.named_modules, a simple strategy is taking the modules in models.named_modules and not sharing scale for QKV. The current code have several issues. Please refine it later